File: scenarios/web-console/api/session_activity.go
package main

import (
	"context"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"web-console/backends"
	"web-console/internal/backend"
	"web-console/internal/sessionstore"
	"web-console/session"
)

// DOC: docs/internal/MESSAGES-VIEW-PROJECTION-UX.md#session-activity-contract

// sessionActivityPayload is the payload of a HubKindSessionActivity envelope.
// snake_case, like every hub payload.
type sessionActivityPayload struct {
	State          string                `json:"state"`
	Source         string                `json:"source"`
	Confidence     float32               `json:"confidence"`
	Since          string                `json:"since,omitempty"`
	LastOutputAt   string                `json:"last_output_at,omitempty"`
	Prompt         *pendingPromptPayload `json:"prompt,omitempty"`
	Harness        string                `json:"harness,omitempty"`
	HarnessVersion string                `json:"harness_version,omitempty"`
}

type pendingPromptPayload struct {
	Kind         string                `json:"kind"`
	Text         string                `json:"text"`
	Options      []promptOptionPayload `json:"options"`
	Answerable   bool                  `json:"answerable"`
	FreeTextHint string                `json:"free_text_hint,omitempty"`
	Hash         string                `json:"hash,omitempty"`
	Cancellable  bool                  `json:"cancellable,omitempty"`
}

type promptOptionPayload struct {
	Key      string `json:"key"`
	Label    string `json:"label"`
	Selected bool   `json:"selected"`
}

func newSessionActivityPayload(a session.Activity) sessionActivityPayload {
	payload := sessionActivityPayload{
		State:          string(a.State),
		Source:         string(a.Source),
		Confidence:     a.Confidence,
		Since:          formatActivityTime(a.Since),
		LastOutputAt:   formatActivityTime(a.LastOutputAt),
		Harness:        a.Harness,
		HarnessVersion: a.HarnessVersion,
	}
	if a.Prompt != nil {
		prompt := &pendingPromptPayload{
			Kind: a.Prompt.Kind, Text: a.Prompt.Text, FreeTextHint: a.Prompt.FreeTextHint, Options: []promptOptionPayload{},
			Answerable: a.Answerable, Hash: backend.PromptHash(a.Prompt), Cancellable: a.Prompt.Cancellable,
		}
		for _, option := range a.Prompt.Options {
			prompt.Options = append(prompt.Options, promptOptionPayload{Key: option.Key, Label: option.Label, Selected: option.Selected})
		}
		payload.Prompt = prompt
	}
	return payload
}

func formatActivityTime(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.UTC().Format(time.RFC3339Nano)
}

// newActivityDetector builds a session's activity detector: it reads the
// session's screen with the detector for its agent harness and publishes
// changes on the conversation hub.
func (s *Server) newActivityDetector(sess *session.Session) *session.ActivityDetector {
	id := sess.ID
	return session.NewActivityDetector(session.ActivityConfig{
		SessionID: id,
		Screen:    sess.ActivityScreen,
		Harness: func() (string, backend.PromptDetector) {
			agent := s.agentTypes.lookup(id)
			return agent, backends.PromptDetectorFor(agent)
		},
		Publish: func(a session.Activity) {
			s.hub.Publish(HubEnvelope{SessionID: id, Kind: HubKindSessionActivity, Payload: newSessionActivityPayload(a)})
		},
		Answering: s.promptAnswering.decide,
	})
}

// activityDetector returns a live session's activity detector, or nil.
func (s *Server) activityDetector(sessionID string) *session.ActivityDetector {
	if sessionID == "" || s.sessions == nil {
		return nil
	}
	sess, ok := s.sessions.Get(sessionID)
	if !ok || sess == nil {
		return nil
	}
	return sess.ActivityDetector()
}

// launchHarnesses are the agent executables a launch command can name.
var launchHarnesses = map[string]sessionstore.Agent{
	"claude":   sessionstore.AgentClaude,
	"codex":    sessionstore.AgentCodex,
	"opencode": sessionstore.AgentOpenCode,
	"grok":     sessionstore.AgentGrok,
}

// agentFromLaunchCommand names the harness a launch command runs: the first
// word whose executable name is a known agent ("/…/shims/claude --model x" is
// claude). Empty when none is named.
func agentFromLaunchCommand(command string) string {
	for _, word := range strings.Fields(command) {
		word = strings.Trim(word, `'"`)
		if agent, ok := launchHarnesses[filepath.Base(word)]; ok {
			return string(agent)
		}
	}
	return ""
}

// agentTypeCacheTTL bounds how stale a session's cached agent type can be; a
// session's agent becomes known at launch or on its first hook.
const agentTypeCacheTTL = 10 * time.Second

// agentTypeCache answers "which harness runs in this session" for detector
// selection without a store read per screen evaluation.
type agentTypeCache struct {
	server *Server
	mu     sync.Mutex
	byID   map[string]agentTypeEntry
}

type agentTypeEntry struct {
	agent string
	at    time.Time
}

func (c *agentTypeCache) lookup(sessionID string) string {
	c.mu.Lock()
	if entry, ok := c.byID[sessionID]; ok && time.Since(entry.at) < agentTypeCacheTTL {
		c.mu.Unlock()
		return entry.agent
	}
	c.mu.Unlock()
	agent := ""
	if c.server != nil && c.server.sessionStore != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		meta, err := c.server.sessionStore.Get(ctx, sessionID)
		cancel()
		if err != nil {
			// Not cached: during the startup storm a read can time out, and
			// caching "no harness" left static screens (a waiting prompt)
			// unread until the next frame after the TTL. Retry next time.
			return ""
		}
		agent = string(meta.AgentType)
		if agent == "" || agent == string(sessionstore.AgentNone) {
			// Panes launched by other scenarios (agent-manager) record no
			// agent type; their launch command still names the harness.
			agent = agentFromLaunchCommand(meta.LaunchCommand)
		}
	}
	c.mu.Lock()
	if c.byID == nil {
		c.byID = make(map[string]agentTypeEntry)
	}
	c.byID[sessionID] = agentTypeEntry{agent: agent, at: time.Now()}
	c.mu.Unlock()
	return agent
}

File: scenarios/web-console/ui/src/api/sessionActivity.ts
import {
  SessionActivitySource,
  SessionActivityState,
  type SessionActivity,
} from "@vrooli/proto-types/web-console/v1/sessions/sessions_pb";

// DOC: docs/internal/MESSAGES-VIEW-PROJECTION-UX.md#session-activity-contract

/** What a session's agent is doing between messages, as the server reads it. */
export type ActivityState = "working" | "idle" | "waiting" | "unknown";

/** The evidence behind an activity. */
export type ActivitySource = "screen" | "output_clock" | "hook" | "harness_event";

export interface PromptOptionView {
  key: string;
  label: string;
  selected: boolean;
}

export interface PendingPromptView {
  kind: string;
  /** Empty when the prompt was detected but its content could not be read. */
  text: string;
  options: PromptOptionView[];
  answerable: boolean;
  freeTextHint?: string;
  /** Identifies the prompt for an answer (kind, text, options; not the selection). */
  hash?: string;
  /** The harness lets Escape dismiss it. */
  cancellable?: boolean;
}

export interface SessionActivityView {
  state: ActivityState;
  source: ActivitySource;
  confidence: number;
  /** RFC 3339; when the current state began. */
  since: string;
  lastOutputAt?: string;
  prompt?: PendingPromptView;
  harness: string;
  harnessVersion?: string;
}

/** The snake_case payload of a `session_activity` hub envelope. */
export interface SessionActivityPayload {
  state?: string;
  source?: string;
  confidence?: number;
  since?: string;
  last_output_at?: string;
  prompt?: {
    kind?: string;
    text?: string;
    options?: { key?: string; label?: string; selected?: boolean }[];
    answerable?: boolean;
    free_text_hint?: string;
    hash?: string;
    cancellable?: boolean;
  };
  harness?: string;
  harness_version?: string;
}

const STATES: ReadonlySet<string> = new Set<ActivityState>(["working", "idle", "waiting", "unknown"]);
const SOURCES: ReadonlySet<string> = new Set<ActivitySource>(["screen", "output_clock", "hook", "harness_event"]);

function asState(value: string | undefined): ActivityState {
  return value && STATES.has(value) ? (value as ActivityState) : "unknown";
}

function asSource(value: string | undefined): ActivitySource {
  return value && SOURCES.has(value) ? (value as ActivitySource) : "output_clock";
}

export function decodeActivityPayload(payload: SessionActivityPayload): SessionActivityView {
  const prompt = payload.prompt;
  return {
    state: asState(payload.state),
    source: asSource(payload.source),
    confidence: payload.confidence ?? 0,
    since: payload.since ?? "",
    ...(payload.last_output_at ? { lastOutputAt: payload.last_output_at } : {}),
    ...(prompt ? {
      prompt: {
        kind: prompt.kind ?? "unknown",
        text: prompt.text ?? "",
        options: (prompt.options ?? []).map((option) => ({
          key: option.key ?? "",
          label: option.label ?? "",
          selected: option.selected ?? false,
        })),
        answerable: prompt.answerable ?? false,
        ...(prompt.free_text_hint ? { freeTextHint: prompt.free_text_hint } : {}),
        ...(prompt.hash ? { hash: prompt.hash } : {}),
        ...(prompt.cancellable ? { cancellable: true } : {}),
      },
    } : {}),
    harness: payload.harness ?? "",
    ...(payload.harness_version ? { harnessVersion: payload.harness_version } : {}),
  };
}

const PROTO_STATES: Readonly<Record<number, ActivityState>> = {
  [SessionActivityState.WORKING]: "working",
  [SessionActivityState.IDLE]: "idle",
  [SessionActivityState.WAITING]: "waiting",
};

const PROTO_SOURCES: Readonly<Record<number, ActivitySource>> = {
  [SessionActivitySource.SCREEN]: "screen",
  [SessionActivitySource.OUTPUT_CLOCK]: "output_clock",
  [SessionActivitySource.HOOK]: "hook",
  [SessionActivitySource.HARNESS_EVENT]: "harness_event",
};

/** Decode `Session.activity` from the sessions list. */
export function decodeActivityProto(activity: SessionActivity): SessionActivityView {
  return decodeActivityPayload({
    state: PROTO_STATES[activity.state] ?? "unknown",
    source: PROTO_SOURCES[activity.source] ?? "output_clock",
    confidence: activity.confidence,
    since: activity.since,
    last_output_at: activity.lastOutputAt,
    ...(activity.prompt ? {
      prompt: {
        kind: activity.prompt.kind,
        text: activity.prompt.text,
        options: activity.prompt.options.map((option) => ({ key: option.key, label: option.label, selected: option.selected })),
        answerable: activity.prompt.answerable,
        free_text_hint: activity.prompt.freeTextHint,
        hash: activity.prompt.hash,
        cancellable: activity.prompt.cancellable,
      },
    } : {}),
    harness: activity.harness,
    harness_version: activity.harnessVersion,
  });
}
